File: Sjaize/Exhaustive Search/소수 찾기.java
// https://school.programmers.co.kr/learn/courses/30/lessons/42839
import java.util.Arrays;
import java.util.Set;
import java.util.HashSet;

class Solution {
    private static Set<Integer> primes = new HashSet<>(); 
    
    private boolean isPrime(int num) {
        if (num <= 1) return false;
        for (int i = 2; i * i <= num; i++) {
            if (num % i == 0) return false;
        }
        return true;
    }
    
    private void getPrimes(int acc, int[] numSet, boolean[] isUsed) {
        if (isPrime(acc)) primes.add(acc);
        
        for (int i = 0; i < numSet.length; i++) {
            if (isUsed[i]) continue;
            int nextAcc = acc * 10 + numSet[i];
            isUsed[i] = true;
            getPrimes(nextAcc, numSet, isUsed);
            isUsed[i] = false;
        }
    }
    
    public int solution(String numbers) {
        int[] numSet = numbers.chars().map(c -> c-'0').toArray();
        boolean[] isUsed = new boolean[numSet.length];
        getPrimes(0, numSet, isUsed);
        return primes.size();
    }
}

File: hall_of_fame/wi-seong-cheol/PGS/Resolve/소수 찾기.cpp
//
//  소수 찾기.cpp
//  main
//
//  Created by wi_seong on 3/21/24.
//

// 시간복잡도: O(nlogn)
// 최악시간: 70,000,000
// 난이도: Level 2
// Timer: 20m
// Url: https://school.programmers.co.kr/learn/courses/30/lessons/42839

#include <iostream>
#include <vector>
#include <unordered_set>
#include <algorithm>

using namespace std;
const int MXN = 10000001;
vector<bool> isPrime(MXN);
unordered_set<int> s;
bool isused[7];

void seive() {
    isPrime[0] = isPrime[1] = true;
    for(int i = 2; i * i < MXN; i++) {
        if(isPrime[i]) continue;
        for(int j = i + i; j <= MXN; j+=i) {
            isPrime[j] = true;
        }
    }
}

void solve(int k, int len, string cur, string numbers) {
    if(k == len) {
        int num = stoi(cur);
        if(!isPrime[num]) {
            s.insert(num);
        }
        return;
    }

    for(int i = 0; i < int(numbers.size()); i++) {
        if(isused[i]) continue;
        isused[i] = true;
        solve(k + 1, len, cur + numbers[i], numbers);
        isused[i] = false;
    }
}

int solution(string numbers) {
    int answer = 0;
    seive();
    int n = int(numbers.size());

    for(int len = 1; len <= n; len++) {
        solve(0, len, "", numbers);
    }
    answer = int(s.size());

    return answer;
}
